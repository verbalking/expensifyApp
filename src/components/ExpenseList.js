
import {connect} from 'react-redux';
import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) =>(
    <div>
        {props.expenses.length === 0 ? (
            <p>No Expenses :(</p>
        ) : (
            
            props.expenses.map((expense)=>{
            return <ExpenseListItem key={expense.id} {...expense}/>;
        })
        )}

    </div>
);

//rebranding this component so as to connect it with the store
const mapStateToProps = (state)=>{
    return{
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

