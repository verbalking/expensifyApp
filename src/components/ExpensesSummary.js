import React from'react';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

import {connect} from 'react-redux';

const ExpensesSummary = ({expenses, total}) =>{

    return(
      <div>
        <h3> Viewing {expenses.length} expenses totalling {total}</h3>
      </div>
    );

} 

const mapStateToProps = (state, props) =>{
  const expensesSelected = getVisibleExpenses(state.expenses, state.filters);
  return  {
      expenses: expensesSelected ,
      total: getExpensesTotal(expensesSelected)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);