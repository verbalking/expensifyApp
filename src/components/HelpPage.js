import React from 'react'

const HelpPage = () => (
    <div>
       This if from Help page
    </div>
);

export default HelpPage;