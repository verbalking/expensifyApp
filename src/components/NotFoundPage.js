import React from 'react'
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom'

const NotFoundPage = () => (
    <div>
       404! - <Link to="/">Go Home you're drunk</Link>
    </div>
);

export default NotFoundPage;