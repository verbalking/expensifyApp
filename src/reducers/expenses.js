const expensesReducerDState=[];


export default (state = expensesReducerDState, action) =>{
    switch(action.type){
        case 'ADD_EXPENSE':
            return [...state, action.expense];  

        case 'REMOVE_EXPENSE':
            const array=state.filter(({id})=> action.id !== id   );
            return array;

        case 'EDIT_EXPENSE':
            return state.map((expense)=>{
                if (expense.id===action.id){
                    return {
                        ...expense,
                        ...action.updates
                    };
                }else{
                    return expense;
                }
            })
            
        default:
            return state;
    }
};