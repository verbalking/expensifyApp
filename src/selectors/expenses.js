import moment from 'moment';


export default (expenses, {text, sortBy, startDate, endDate})=>{
    return expenses.filter((expense)=>{
        const createdAtMoment= moment(expense.createdAt);
        const startDateMatch= startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') :true;  
        const endDateMatch= endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') :true;

        const description=expense.description.toLowerCase();
        const textMatch= description.includes(text.toLowerCase()); 
        return startDateMatch && endDateMatch && textMatch;
        //filter only returns the objects that satisfy true in the return()
        
    }).sort((a, b)=>{
        if(sortBy==='date'){
            return a.createdAt>b.createdAt ? 1:-1;
        }else if(sortBy==='amount'){
            return a.amount>b.amount ? 1:-1;
        }
    });
};