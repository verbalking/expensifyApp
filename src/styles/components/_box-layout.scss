.box-layout{
  background: url('/images/bg.jpg');
  background-size: cover;
  height: 100vh;
  width: 100vw;
  display: flex;
  align-items: center ;
  justify-content: center;
}

.box-layout__box{
  background: fade-out(white, .30);
  border-radius: 3px;
  padding: $l-size $m-size; //inside of the box
  text-align: center;
  width: 25rem;

}

.box-layout__title{
  margin: 0 0 $m-size 0;
  line-height: 1;
}