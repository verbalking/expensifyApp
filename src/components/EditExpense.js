import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startRemoveExpense, startEditExpense} from '../actions/expenses'


export class EditExpense extends React.Component{

    onSubmit=(expense)=>{
    
        this.props.startEditExpense(this.props.expense.id ,expense);
        this.props.history.push('/');
    };

    render(){
        return(
            <div>
            <ExpenseForm 
                expense={this.props.expense}
                onSubmit={this.onSubmit}
            />
    
            <button onClick={()=>{
                this.props.startRemoveExpense({id: this.props.expense.id});
                this.props.history.push('/');
            }}>Remove </button>
    
        </div>
        )
    };
}

const mapStateToProps = (state, props) =>{
    return  {
        expense: state.expenses.find( (expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps=(dispatch)=>({
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);