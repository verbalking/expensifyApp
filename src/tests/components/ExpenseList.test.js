import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseList} from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';


test ('should render expense list ', ()=>{
  const wrapper = shallow(<ExpenseList expenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
});

test ('should render expense list empty ', ()=>{
  const wrapper = shallow(<ExpenseList expenses={[]}/>);
  expect(wrapper).toMatchSnapshot();
});
