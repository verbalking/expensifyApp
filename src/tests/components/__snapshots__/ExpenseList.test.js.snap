// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list  1`] = `
<div>
  <ExpenseListItem
    amount={10}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={20}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={30}
    createdAt={345600000}
    description="Credit card"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render expense list empty  1`] = `
<div>
  <p>
    No Expenses :(
  </p>
</div>
`;
